monitor-function-calls/src/agent.spec.js

```diff
@@ -0,0 +1,621 @@
+            // these delete statements will still work even if the keys don't exist
+            /* eslint-disable no-param-reassign */
+            delete functionSignature.expression;
+            delete functionSignature.expressionObject;
+            /* eslint-enable no-param-reassign */
+          }
+        });
+      });
+
+      // run the handler
+      const findings = await handleTransaction(mockTxEvent);
+
+      const argumentData = {};
+      argNames.forEach((name, argIndex) => {
+        if (argumentData[name] == null) {
+          argumentData[name] = argValues[argIndex].toString();
+        }
+      });
+
+      // create the expected finding
+      const testFindings = [Finding.fromObject({
+        alertId: `${config.developerAbbreviation}-${config.protocolAbbreviation}-FUNCTION-CALL`,
+        description: `The ${testConfig.functionInConfig.name} function was invoked in the ${contractName} contract`,
+        name: `${config.protocolName} Function Call`,
+        protocol: config.protocolName,
+        severity: FindingSeverity[testConfig.findingSeverity],
+        type: FindingType[testConfig.findingType],
+        metadata: {
+          contractAddress: validContractAddress,
+          contractName,
+          functionName: testConfig.functionInConfig.name,
+          ...argumentData,
+        },
+      })];
+
+      expect(findings).toStrictEqual(testFindings);
+    });
+
+    it('returns a finding if a target contract invokes a monitored function when the expression condition is met', async () => {
+      let expression;
+      let expressionObject;
+      initializeData.contracts.forEach((contract) => {
+        const { functionSignatures } = contract;
+        functionSignatures.forEach((functionSignature) => {
+          if (functionSignature.functionName === testConfig.functionInConfig.name) {
+            expressionObject = functionSignature.expressionObject;
+            expression = functionSignature.expression;
+          }
+        });
+      });
+      const { variableName, operator, value } = expressionObject;
+
+      // encode function data
+      // valid function name with valid arguments
+      const { argValues, argNames } = createMockArgs(testConfig.functionInConfig);
+
+      // override the argument value that corresponds to the expression condition
+      const argOverride = { name: variableName };
+      if (BigNumber.isBigNumber(value)) {
+        switch (operator) {
+          case '>=':
+          case '<=':
+          case '===':
+            argOverride.value = value.toString();
+            break;
+          case '>':
+          case '!==':
+            argOverride.value = value.plus(1).toString();
+            break;
+          case '<':
+            argOverride.value = value.minus(1).toString();
+            break;
+          default:
+            throw new Error(`Unknown operator: ${operator}`);
+        }
+      } else if (typeof (value) === 'string') {
```

Comment:
This `else if` block should be below the one that checks if a value is a hex string, since otherwise we'll never reach that path since all hex strings are also JS string types